spinand.c

```diff
@@ -0,0 +1,279 @@
+/*
+ * This file is part of the flashrom project.
+ *
+ * Copyright (C) 2018 Yangfl
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 2 of the License.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+/*
+ * Contains the SPI NAND chip driver functions
+ */
+
+#include <stddef.h>
+#include <string.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include "flash.h"
+#include "flashchips.h"
+#include "chipdrivers.h"
+#include "programmer.h"
+#include "spi.h"
+#include "spinand.h"
+
+/***************************************
+ * UTIL FUNCTIONS
+ ***************************************/
+
+/* For error number */
+static int errno;
+
+/* 1 dummy byte between input and output */
+#define DUMMY_BYTE 1
+
+static int spi_nand_prepare_row_address(uint8_t cmd_buf[], const unsigned int addr)
+{
+	cmd_buf[1] = (addr >> 16) & 0xff;
+	cmd_buf[2] = (addr >>  8) & 0xff;
+	cmd_buf[3] = (addr >>  0) & 0xff;
+	return 3;
+}
+
+static int spi_nand_prepare_column_address(uint8_t cmd_buf[], const unsigned int addr)
+{
+	cmd_buf[1] = (addr >>  8) & 0xff;
+	cmd_buf[2] = (addr >>  0) & 0xff;
+	return 2;
+}
+
+/***************************************
+ * CHIP FEATURE
+ ***************************************/
+
+uint8_t spi_nand_get_feature_multi(struct flashctx *flash, unsigned char addr, int time)
+{
+	/* Feature will be outputted continuously until CS goes high */
+	unsigned char cmd[SPINAND_GET_FEATURE_OUTSIZE] = { SPINAND_GET_FEATURE, addr };
+	int ret;
+	unsigned char readarr[time];
+
+	if (time > 100) {
+		msg_cerr("%s called with too many 'time'! Please report a bug at "
+			 "flashrom@flashrom.org\n", __func__);
+		errno = SPI_FLASHROM_BUG;
+		return 0;
+	}
+
+	ret = spi_send_command(flash, sizeof(cmd), sizeof(readarr), cmd, readarr);
+	if (ret) {
+		msg_cerr("GET FEATURE failed!\n");
+		errno = ret;
+		return 0;
+	}
+
+	msg_cspew("GET FEATURE 0x%01x returned 0x%01x. ", addr, readarr[time - 1]);
+	return readarr[time - 1];
```

Comment:
This seems wrong. I don't understand what `time` has to do with a command response.